service/src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java

```diff
@@ -0,0 +1,113 @@
+                pageRequestDTO.getPage(), pageRequestDTO.getSize());
+        List<OrderDTO> orderDTOList = converter.convertToListDTO(orderList);
+        int totalElements = orderRepository.countEntities(specification);
+        return new PageDTO<>(
+                pageRequestDTO.getPage(),
+                pageRequestDTO.getSize(),
+                totalElements,
+                orderDTOList
+        );
+    }
+
+    @Override
+    public OrderDTO create(Long userId, OrderDTO orderDTO) {
+        User user = getUserIfExists(userId);
+        Order order = converter.convertToEntity(orderDTO);
+        order.setUser(user);
+        List<Certificate> fullCertificates = getFullCertificates(order);
+        order.setCertificates(fullCertificates);
+        BigDecimal totalPrice = countTotalPrice(order);
+        order.setTotalPrice(totalPrice);
+        return converter.convertToDto(
+                orderRepository.save(order)
+        );
+    }
+
+    @Override
+    public PageDTO<OrderDTO> getAllUserOrders(Long userId, PageRequestDTO pageRequestDTO) {
+        User user = getUserIfExists(userId);
+        CriteriaSpecification<Order> specification = new AllUserOrdersSpecification(user);
+        List<Order> userOrderList = orderRepository.getEntityListBySpecification(specification,
+                pageRequestDTO.getPage(), pageRequestDTO.getSize());
+        List<OrderDTO> userOrderDTOList = converter.convertToListDTO(userOrderList);
+        int totalElements = orderRepository.countEntities(specification);
+        return new PageDTO<>(
+                pageRequestDTO.getPage(),
+                pageRequestDTO.getSize(),
+                totalElements,
+                userOrderDTOList
+        );
+    }
+
+    @Override
+    public OrderDTO getUserOrder(Long userId, Long orderId, PageRequestDTO pageRequestDTO) {
+        User user = getUserIfExists(userId);
+        CriteriaSpecification<Order> specification = new UserOrderSpecification(user, orderId);
+        Order order = orderRepository.getEntityBySpecification(specification)
+                .orElseThrow(() -> new EntityNotFoundException(" (orderId = " + orderId + ")"));
+        return converter.convertToDto(order);
+    }
+
+    private User getUserIfExists(Long userId) {
+        return userRepository.getById(userId)
+                .orElseThrow(() -> new EntityNotFoundException(" (userId = " + userId + ")"));
+    }
+
+    private List<Certificate> getFullCertificates(Order order) {
+        List<Certificate> fullCertificates = new ArrayList<>();
+        for (Certificate certificate : order.getCertificates()) {
+            Certificate fullCertificate = certificateRepository.getById(certificate.getId())
+                    .orElseThrow(() -> new EntityNotFoundException(" (certificateId = " + certificate.getId() + ")"));
+            fullCertificates.add(fullCertificate);
+        }
+        return fullCertificates;
```

Comment:
You could replace this block with:
        return order.getCertificates().stream()
                .map(certificate -> certificateRepository.getById(certificate.getId())
                        .orElseThrow(() -> new EntityNotFoundException(" (certificateId = " + certificate.getId() + ")")))
                .collect(Collectors.toList());

Reply:
Thanks, Very cool solution, I like it